desmod/simulation.py

```diff
@@ -183,6 +183,7 @@ def simulate(config, top_type, env_type=SimEnvironment, reraise=True,
                 except BaseException as e:
                     env.tracemgr.trace_exception()
                     result['sim.exception'] = repr(e)
+                    top.post_fail()
```

Comment:
Thanks for pointing out the possibility of `top` being unavailable here.  

Yes, you are right in the sense that anyone writing a `Component` with a `post_fail_hook()` should be aware that a valid `Component` exists for `Component.post_fail_hook()` to be called. I don't know if that's a big deal. Another possibility is to use have a `Component.post_fail_hook()` and `SimEnvironment.post_fail_hook()` to handle the case where `Component` is invalid. I'm not sure if that's necessary.

I don't understand what you mean by tear-down any _resources_ that might have been affected by `post_simulate()`. Can you elaborate on what you mean by _resources_?